forge/src/debugger.rs

```diff
@@ -0,0 +1,212 @@
+#[derive(Default, Debug, Clone)]
+pub struct DebugNode {
+    /// Parent node index in the arena
+    pub parent: Option<usize>,
+    /// Children node indexes in the arena
+    pub children: Vec<usize>,
+    /// Location in parent
+    pub location: usize,
+    /// Address context
+    pub address: Address,
+    /// Depth
+    pub depth: usize,
+    /// The debug steps
+    pub steps: Vec<DebugStep>,
+    /// Whether the contract was created in this node or not
+    pub creation: bool,
+}
+
+impl DebugNode {
+    pub fn new(address: Address, depth: usize, steps: Vec<DebugStep>) -> Self {
+        Self { address, depth, steps, ..Default::default() }
+    }
+}
+
+/// A `DebugStep` is a snapshot of the EVM's runtime state.
+///
+/// It holds the current program counter (where in the program you are),
+/// the stack and memory (prior to the opcodes execution), any bytes to be
+/// pushed onto the stack, and the instruction counter for use with sourcemap.
+#[derive(Debug, Clone)]
+pub struct DebugStep {
+    /// The program counter
+    pub pc: usize,
+    /// Stack *prior* to running the associated opcode
+    pub stack: Vec<U256>,
+    /// Memory *prior* to running the associated opcode
+    pub memory: Memory,
+    /// Opcode to be executed
+    pub instruction: Instruction,
+    /// Optional bytes that are being pushed onto the stack
+    pub push_bytes: Option<Vec<u8>>,
+    /// Instruction counter, used to map this instruction to source code
+    pub ic: usize,
+    /// Cumulative gas usage
+    pub total_gas_used: u64,
+}
+
+impl Default for DebugStep {
+    fn default() -> Self {
+        Self {
+            pc: 0,
+            stack: vec![],
+            memory: Memory::new(),
+            instruction: Instruction::OpCode(revm::opcode::INVALID),
+            push_bytes: None,
+            ic: 0,
+            total_gas_used: 0,
+        }
+    }
+}
+
+impl DebugStep {
+    /// Pretty print the step's opcode
+    pub fn pretty_opcode(&self) -> String {
+        if let Some(push_bytes) = &self.push_bytes {
+            format!("{}(0x{})", self.instruction, hex::encode(push_bytes))
+        } else {
+            self.instruction.to_string()
+        }
+    }
+}
+
+impl Display for DebugStep {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some(push_bytes) = &self.push_bytes {
+            write!(
+                f,
+                "pc: {:?}\nop: {}(0x{})\nstack: {:#?}\nmemory: 0x{}\n\n",
+                self.pc,
+                self.instruction,
+                hex::encode(push_bytes),
+                self.stack,
+                hex::encode(self.memory.data())
+            )
+        } else {
+            write!(
+                f,
+                "pc: {:?}\nop: {}\nstack: {:#?}\nmemory: 0x{}\n\n",
+                self.pc,
+                self.instruction,
+                self.stack,
+                hex::encode(self.memory.data())
+            )
+        }
+    }
+}
```

Comment:
This is not actually used anywhere, so we can remove it unless someone wants to use it :smile:

Reply:
sgtm